### DISCUSSION TITLE
Time-series for data streams

### DISCUSSION THREAD

**Discussion by cvijayadurga** [online-ml/river#987] [Q&A] [answered]
Hi All,

Just a general question here
In the case of pre-existing datasets such as the aeroplane dataset, concepts of trend, seasonality can be verified. Just wondering if these concepts still apply for data streams, since we are only concerned with data points at a given time, how do the concepts of seasonality and trend apply time-series predictions for data streams?

**Comment by MaxHalford** [comment, marked as answer]
Hey hey.

> aeroplane dataset

Do you mean the `AirlinePassengers` dataset?

> Just wondering if these concepts still apply for data streams, since we are only concerned with data points at a given time, how do the concepts of seasonality and trend apply time-series predictions for data streams?

I'm not sure I fully understand the question, but here's an answer anyway. Time series datasets are special in that the order the data arrives matters, and as you the target usually exhibits some kind of trend and/or seasonality. Nearly all time series models try to model this internally. But I'm not sure exactly what you're interested in. Could you expand?

**Reply by cvijayadurga** [reply]
Apologies for the ambiguous wording. I did mean the 'AirlinePassengers' dataset. To put it another way, let's consider any real-time sensor data stream capturing one data point at a time. How are trends and seasonality of such data identified given we don't have any historical references? are these concepts applicable when we are dealing with just one data point?

**Reply by MaxHalford** [reply]
> Apologies for the ambiguous wording

It's no problem, I'm just trying to make sure I understand :)

> How are trends and seasonality of such data identified given we don't have any historical references?

It really depends on the model's internals. For instance, in SNARIMAX, past values are used as features. At first there are no past values, so these features are simply not included.

**Comment by cvijayadurga** [comment]
Thank you. That answers my question - I went through the example listed in the nowcasting example(using linear regression) and SNARIMAX and noticed how, in the SNARIMAX  model, there is no mechanism in place to capture trends as done in the nowcasting example.

This brings me to my next question: how are values for p,q, and d determined in the case of SNARIMAX, assuming there is no historical data (like sensor data). I do know that one way to find values of p and q  using PACF and ACF. Do you have any suggestions on determining these values for data streams? 

P.S: `metric = utils.Rolling(metrics.MAE(), 12)` in snarimax.py returned an error. I used Rolling method from metrics instead  as `metric = metrics.Rolling(metrics.MAE(), 12)` for the sample example provided in snarimax.py

**Reply by MaxHalford** [reply]
> in the SNARIMAX model, there is no mechanism in place to capture trends as done in the nowcasting example

Actually there is, but it's subtle. It happens through the differencing mechanism. The model is trains and make predictions on (seasonally) differenced values, rather than the raw values.

> This brings me to my next question: how are values for p,q, and d determined in the case of SNARIMAX, assuming there is no historical data (like sensor data). I do know that one way to find values of p and q using PACF and ACF. Do you have any suggestions on determining these values for data streams?

I don't have a good answer. Having an online way to do that would be amazing :). The brute-force way would be to have multiple copies of the model with different parameters running in parallel. Time series forecasting is usually not a computationally intensive task, so maybe it's not such a bad idea.

> metric = utils.Rolling(metrics.MAE(), 12)

That should work in River's next release :)

**Reply by cvijayadurga** [reply]
I did use brute force to estimate p,q, and d  and chose the values yielding the least MAE but was wondering if there was a better mechanism. Nevertheless, thank you for fruitful discussion.

**Reply by dberardo-com** [reply]
would it be possible to change the seasonal parameters of an existing SNARIMAX model without having to reinitialize it? So basically resetting only the sasonality related params, without loosing information on the other params? @MaxHalford 

This way one might attempt some sort of online correction of seasonal params.

> The brute-force way would be to have multiple copies of the model with different parameters running in parallel.

Would this be some sort of dynamic "model selection" pipeline that picks models based on the performance metric score?

> I did use brute force to estimate p,q, and d and chose the values yielding the least MAE but was wondering if there was a better mechanism. Nevertheless, thank you for fruitful discussion.

@brownbytes  have you thought of using fourier transforms analysis? River implements an online version of it, and i have a gut feeling that seasonality information should somehow be available in the signal specturm. here some more information (although for an anomalies detection use case: https://github.com/online-ml/river/discussions/978#discussioncomment-3077212)

**Reply by MaxHalford** [reply]
> would it be possible to change the seasonal parameters of an existing SNARIMAX model without having to reinitialize it? So basically resetting only the sasonality related params, without loosing information on the other params? @MaxHalford

Yes. In fact, the new release will make this a smooth process. See [here](https://riverml.xyz/dev/recipes/cloning-and-mutating/#mutating-attributes).

> Would this be some sort of dynamic "model selection" pipeline that picks models based on the performance metric score?

Yes exactly. It's online model selection :)

**Reply by cvijayadurga** [reply]
@dberardo-com : nope, I did not give Fourier transforms a try. Thanks for the suggestion though :)

**Reply by dberardo-com** [reply]
let me know if you get any useful results using fourier, as i am still experimenting with it.

**Comment by cvijayadurga** [comment]
@MaxHalford  - Could you please help me understand how is the paper "Online Learning for Time Series Prediction" used for TS in river? I tried reading and putting it in perspective but just couldn't.
Reason I ask is - as I understand, SNARIMAX module operates very much like regression in batch learning - build features using TS parameters (p,d,q,...) ,normalise features(using moving averages unlike batch learning) -> feed to regression model -> make a forecast ->calculate loss  & update weights 
So where are any results ( if any ) from that paper used?

**Reply by MaxHalford** [reply]
Hey there. I cited that paper because of the idea of using a linear regression (or any kind of model). Indeed, normally a Kalman filter is used to learn the parameters of ARIMA models. So no there aren't results per say, it's just the idea that mattered to me when I read the paper.

**Reply by cvijayadurga** [reply]
I see. Thank you!
